caas/kubernetes/provider/bootstrap.go

```diff
@@ -307,15 +308,29 @@ func (c *controllerStack) doCleanUp() {
 	}
 }
 
-// Deploy creates all resources for controller stack.
-func (c *controllerStack) Deploy() (err error) {
+// Deploy creates all resources for the controller stack.
+func (c *controllerStack) Deploy(ctx context.Context) (err error) {
 	// creating namespace for controller stack, this namespace will be removed by broker.DestroyController if bootstrap failed.
 	nsName := c.broker.GetCurrentNamespace()
 	c.ctx.Infof("Creating k8s resources for controller %q", nsName)
 	if err = c.broker.createNamespace(nsName); err != nil {
 		return errors.Annotate(err, "creating namespace for controller stack")
 	}
 
+	// The given context is checked for cancellation manually between each
+	// step (not ideal, but it avoids wiring context absolutely everywhere).
+	isDone := func(ctx context.Context) bool {
+		select {
+		case <-ctx.Done():
+			return true
+		default:
+			return false
+		}
+	}
+	if isDone(ctx) {
+		return errors.Errorf("cancelled")
```

Comment:
Let's define an ErrCancelled to use in all the places

Reply:
Done as `bootstrap.Cancelled()` / `bootstrap.IsCancelled()` in https://github.com/juju/juju/pull/12475